File: python/isbn-verifier/isbn_verifier.py
"""Solution to the ISBN Verifier exercise."""


def is_valid(isbn: str) -> bool:
    """Check whether the given ``isbn`` is valid.

    Args:
        isbn (str): the isbn number to check

    Returns:
        bool: ``True`` if the given ``isbn`` is valid.
    """
    
    digits = [x.replace('X', '10') for x in list(isbn.replace('-', ''))]

    if len(digits) != 10 or any(not d.isdigit() for d in digits) or any(d == '10' for d in digits[:-1]):
        return False

    checksum = sum((10 - i) * int(d) for i, d in enumerate(digits))

    return checksum % 11 == 0

File: kotlin/isbn-verifier/src/main/kotlin/IsbnVerifier.kt
private const val CHECK_DIGIT = 'X'

class IsbnVerifier {

    fun isValid(number: String): Boolean {
        val digits = number.filter { it.isLetterOrDigit() }

        if (digits.length != 10) {
            return false
        }

        var checksum = 0

        for ((index, digit) in digits.reversed().withIndex()) {
            checksum += when {
                digit == CHECK_DIGIT && index == 0 -> 10
                digit.isDigit() -> digit.digitToInt() * (index + 1)
                else -> return false
            }
        }

        return checksum % 11 == 0
    }
}
